[APM] Refactors property tables into single metadata table

jasonrhodes opened the pull request:
Closes #24648

## Summary

Rearranges the property-based tabs into a single "Metadata" tab, organized by section into one list. This change happens on the transaction detail page, the error detail page, and in the transaction flyout that opens when you click on a transaction in the timeline view.

<img width="1371" alt="Screen Shot 2019-04-16 at 7 50 23 AM" src="https://user-images.githubusercontent.com/159370/56207383-97953680-601c-11e9-9a1f-55bfe1ef0661.png">

<img width="1369" alt="Screen Shot 2019-04-16 at 7 50 53 AM" src="https://user-images.githubusercontent.com/159370/56207391-9bc15400-601c-11e9-955b-a00e51a27117.png">

<img width="1170" alt="Screen Shot 2019-04-16 at 7 51 23 AM" src="https://user-images.githubusercontent.com/159370/56207405-a0860800-601c-11e9-98f2-03d70aed256a.png">

jasonrhodes committed:
Refactors property table usage with new components and tests

jasonrhodes committed:
Changes default tab for transaction detail view

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9795/)

formgeist reviewed (changes requested):
Looks good (so far) - I have a few change requests;

- Not sure why it isn't, but the Timeline should still be the default tab when viewing the Transaction detail.
- The Transaction detail flyout doesn't feature the Timeline, so if it's possible change the tabs to a title for Metadata?
- Can we add the "How to..." button to consolidate all the help messages that are at labels, user and custom?
- That 50/50 width is still not configurable, right?
- Why does so many keys have `0` at the end?

jasonrhodes commented:
> * Not sure why it isn't, but the Timeline should still be the default tab when viewing the Transaction detail.

That was fixed here: https://github.com/elastic/kibana/pull/35150/commits/177adc82d6e14608effdae261caab4495b0b0006

> * The Transaction detail flyout doesn't feature the Timeline, so if it's possible change the tabs to a title for Metadata?

We should be able to do this -- are we sure we don't want to add anything else here?

> * Can we add the "How to..." button to consolidate all the help messages that are at labels, user and custom?

Not sure what you mean here?

> * That 50/50 width is still not configurable, right?

Hm I forget now, let me check again.

> * Why does so many keys have `0` at the end?

Any value that's an array has that treatment. I think it's better to display it than to try to hide it with custom logic, even if it's a little noisy...

formgeist commented:
> That was fixed here: 177adc8

Thanks, I must've missed the commit 👍

> We should be able to do this -- are we sure we don't want to add anything else here?

Not at least for the moment

> Not sure what you mean here?

In https://github.com/elastic/kibana/issues/24648 I added a task to add a button in the top to link to a documentation article (to be written) that consolidated the help messages we'd display for labels, user and custom sections to help the user get started with filling those sections.

![01 DT - Transaction detail - Context (show all)](https://user-images.githubusercontent.com/4104278/56211408-270c0b00-6058-11e9-8f82-3b692b69bac9.png)

> Any value that's an array has that treatment. I think it's better to display it than to try to hide it with custom logic, even if it's a little noisy...

OK, just needed to understand why that was happening. I guess it's better to leave as is.

Thanks!

ogupte commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/TransactionTabs.tsx:
```diff
@@ -7,38 +7,38 @@
   key: 'timeline';
   label: string;
 }
 
```
is there any benefit for differentiating the TimelineTab and MetadataTab types? Can we combine them into a single interface `ViewTab` or something like that?

jasonrhodes replied:
Actually now that you mention it, there's no reason to not just let TS infer the types of these altogether. Thanks!

ogupte commented on x-pack/plugins/apm/public/components/shared/DottedKeyValueTable/FormattedValue.tsx:
```diff
@@ -0,0 +1,42 @@
+  if (isObject(value)) {
+    return <pre>{JSON.stringify(value, null, 4)}</pre>;
+  } else if (isBoolean(value) || isNumber(value)) {
+    return <React.Fragment>{String(value)}</React.Fragment>;
```
could remove this case and use `String(value)` invariably in the last return statement.

jasonrhodes replied:
Yeah we could ... I think I'll leave this as-is right now because I'm not 100% sure if there's a reason to not stringify the value unless we have to.

ogupte commented on x-pack/plugins/apm/public/components/shared/Stacktrace/Stackframe.tsx:
```diff
@@ -49,6 +49,7 @@ export function Stackframe({
   isLibraryFrame = false
 }: Props) {
   if (!hasLineContext(stackframe)) {
+    console.log('no line context');
```
is this console.log intentional? if so maybe the messaging should have more context about lack of line context. 😄

jasonrhodes replied:
Woops! Why did the linter let me commit this???

jasonrhodes committed:
Small refactors to property table

jasonrhodes committed:
Review feedback

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9956/)

formgeist reviewed (dismissed):
Looks good again. I think I'm forgetting that we should make the span flyout consistent, so it shows "Metadata" instead of "Labels" and add a "Labels" section in that tab. Would that be possible in this PR?

<img width="1180" alt="Screenshot 2019-04-17 at 09 08 40" src="https://user-images.githubusercontent.com/4104278/56267975-98979800-60f0-11e9-9dc6-15b33feb6e1e.png">

jasonrhodes commented:
@formgeist I addressed all feedback, I think, except for the 50% width thing -- I set the table to "table-layout: auto" which made the columns stop sticking to 50%, but they got really weird and some were big and some were small and it looked awful. I can show you on a call if you like. :)

jasonrhodes committed:
Updates translations

jasonrhodes commented:
@formgeist do we already have a link for that help link from your mockup? Do we need to pull in @bmorelli25 to help with that? I think this PR is now blocked by the existence of that article...

formgeist commented:
@jasonrhodes I already created https://github.com/elastic/apm-server/issues/2111 to setup that documentation article. We're still discussing the approach whether it should be the same for all agents or each agent individually (like we linked the agent documentation previously). Let's see if we can resolve this today, and get the link in later. Otherwise, we'll have to pull it out and add in another PR when the page(s) are ready.

jasonrhodes commented:
> Let's see if we can resolve this today, and get the link in later. Otherwise, we'll have to pull it out and add in another PR when the page(s) are ready.

Sounds great, thanks for the link to that other issue. :+1:

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10014/)

bmorelli25 commented:
@jasonrhodes - The "How to add.." docs link should point to
```js
{config.docs.base_url}guide/en/apm/get-started/{config.docs.version}/apm-data-model.html
```
which should be:
```js
https://www.elastic.co/guide/en/apm/get-started/master/apm-data-model.html
```

More info here: https://github.com/elastic/apm-server/pull/2125

jasonrhodes committed:
Refactors metadata and adds tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10192/)

jasonrhodes committed:
Rearranges documentation links to prefer declarative component

jasonrhodes committed:
Removes unused component

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10227/)

jasonrhodes committed:
Improves metadata component tests and removes giant snapshots

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10243/)
